psci/Main.hs

```diff
@@ -0,0 +1,140 @@
+
+import qualified Paths_purescript as Paths
+
+import           System.Console.Haskeline
+import           System.Exit
+import           System.FilePath.Glob (glob)
+import           System.IO.UTF8 (readUTF8File)
+
+multiLineMode :: Opts.Parser Bool
+multiLineMode = Opts.switch $
+     Opts.long "multi-line-mode"
+  <> Opts.short 'm'
+  <> Opts.help "Run in multi-line mode (use ^D to terminate commands)"
+
+inputFile :: Opts.Parser FilePath
+inputFile = Opts.strArgument $
+     Opts.metavar "FILE"
+  <> Opts.help "Optional .purs files to load on start"
+
+inputForeignFile :: Opts.Parser FilePath
+inputForeignFile = Opts.strOption $
+     Opts.short 'f'
+  <> Opts.long "ffi"
+  <> Opts.help "The input .js file(s) providing foreign import implementations"
+
+nodeFlagsFlag :: Opts.Parser [String]
+nodeFlagsFlag = Opts.option parser $
+     Opts.long "node-opts"
+  <> Opts.metavar "NODE_OPTS"
+  <> Opts.value []
+  <> Opts.help "Flags to pass to node, separated by spaces"
+  where
+    parser = words <$> Opts.str
+
+psciOptions :: Opts.Parser PSCiOptions
+psciOptions = PSCiOptions <$> multiLineMode
+                          <*> many inputFile
+                          <*> many inputForeignFile
+                          <*> nodeFlagsFlag
+
+version :: Opts.Parser (a -> a)
+version = Opts.abortOption (Opts.InfoMsg (showVersion Paths.version)) $
+            Opts.long "version" <>
+            Opts.help "Show the version number" <>
+            Opts.hidden
+
+getOpt :: IO PSCiOptions
+getOpt = Opts.execParser opts
+    where
+      opts        = Opts.info (version <*> Opts.helper <*> psciOptions) infoModList
+      infoModList = Opts.fullDesc <> headerInfo <> footerInfo
+      headerInfo  = Opts.header   "psci - Interactive mode for PureScript"
+      footerInfo  = Opts.footer $ "psci " ++ showVersion Paths.version
+
+-- | Read a file in the 'P.Make' monad, handling errors.
+readFileMake :: FilePath -> P.Make String
+readFileMake path = P.makeIO (const (P.ErrorMessage [] $ P.CannotReadFile path)) (readUTF8File path)
+
+-- | Parses the input and returns either a command, or an error as a 'String'.
+getCommand :: Bool -> InputT (StateT PSCiState IO) (Either String (Maybe Command))
+getCommand singleLineMode = handleInterrupt (return (Right Nothing)) $ do
+  firstLine <- withInterrupt $ getInputLine "> "
+  case firstLine of
+    Nothing -> return (Right (Just QuitPSCi)) -- Ctrl-D when input is empty
+    Just "" -> return (Right Nothing)
+    Just s | singleLineMode || head s == ':' -> return .fmap Just $ parseCommand s
+    Just s -> fmap Just . parseCommand <$> go [s]
+  where
+    go :: [String] -> InputT (StateT PSCiState IO) String
+    go ls = maybe (return . unlines $ reverse ls) (go . (:ls)) =<< getInputLine "  "
+
+-- | Checks if the Console module is defined
+consoleIsDefined :: [P.ExternsFile] -> Bool
+consoleIsDefined = any ((== P.ModuleName (map P.ProperName [ "Control", "Monad", "Eff", "Console" ])) . P.efModuleName)
```

Comment:
Would it make sense to have a pattern synonym for `P.ModuleName (map P.ProperName [ "Control", "Monad", "Eff", "Console" ])`?